### ISSUE TITLE
Introduce PTQWrapper class for calibration + export with multiple PTQ algorithm

### ISSUE THREAD

**Issue by mhs4670go** [Samsung/TICO#89]
### Motivation

Quantizing a pretrained model in PyTorch typically involves hand-crafting per-layer calibration loops, duplicating code for each observer algorithm, and juggling separate state for min/max, histograms, or reconstruction parameters. A unified `PTQWrapper` solves these pain points by:

- Encapsulating each layer’s calibration logic, fake-quant calls, and parameter storage in a single module  
- Allowing multiple observer-based algorithms to collect statistics in one pass and be switched at will for head-to-head comparison  
- Centralizing `state_dict` management so you can checkpoint, resume, and export a fully instrumented model with minimal boilerplate  

By introducing `PTQWrapper`, we streamline end-to-end PTQ workflows, reduce code duplication, and ensure reproducible, exportable quantization pipelines.  

Below is a suggested design for a flexible, custom PTQ wrapper module in PyTorch—complete with considerations, a sketch of the API/architecture, and a concrete code example.

### Design Overview

```
       +-------------------+
       |   FP32 Model      |
       +-------------------+
                 |
                 | prepare(model, configs)
                 v
       +-------------------+
       | Wrapped Model     |
       | (PTQWrapper)      |
       +-------------------+
                 |
                 | run_fn(calibration data)
                 v
       +-------------------+
       | Calibration Pass: |
       | • minmax observer |
       | • percentile obs. |
       +-------------------+
                 |
                 | compute_qparams()
                 v
       +-------------------+
       | QParams Stored    |
       | (scale / zero‐pt) |
       +-------------------+
                 |
                 | set_active_algorithm('minmax')  
                 |   or ('percentile'), etc.
                 v
       +-------------------+
       | Quantized Forward |
       | • fake‐quantize   |
       +-------------------+
                 |
                 | convert(..., algorithm='minmax')
                 v
       +-------------------+
       | Exportable Graph  |
       | (torch.export)    |
       +-------------------+
```

#### PTQWrapper

- **Wraps any** `nn.Module` (e.g. `Conv2d`, `Linear`, etc.)  
- **Holds one or more** Observer/Algorithm objects responsible for:
  - Collecting statistics (min/max, histograms, percentiles, …)  
  - Computing and holding quantization parameters (scale, zero-point, bitwidth)

#### Observer/Algorithm Interface

- **`collect(x)`**  
  - Called during calibration mode to update running stats.  
- **`compute_qparams() → (scale, zero_point)`**  
    - After calibration, computes the final quantization parameters.  
- **`fake_quant(x, scale, zero_point)`**  
    - Applies `torch.fake_quantize_per_tensor_affine` (or per-channel) inside `forward`.

#### Lifecycle & Modes

- **Calibration mode**  
  - `wrapper.enable_calibration()`  
  - Runs the wrapped layer in FP32, but each forward also calls `observer.collect()` on activations (and on weights if desired).
- **Quantization mode**  
  - `wrapper.compute_qparams()`  
  - Optionally `wrapper.freeze_qparams()` (disable further stats updates)  
  - In `forward`:  
    1. Fake-quantize weights  
    2. Compute FP32 output  
    3. Fake-quantize the activation output
- **Export**  
  - When you call `torch.export.export`, because your `forward` calls PyTorch’s fake-quant API, the resulting graph will include those quantize/dequantize ops.

### Key Considerations

- **Per-Tensor vs. Per-Channel**  
  - Decide whether to quantize weight tensors per-channel (common for conv) and activations per-tensor.  
- **Symmetric vs. Asymmetric**  
  - Asymmetric (with zero-point) for activations; symmetric often for weights.  
- **Bitwidth Flexibility**  
  - Allow `num_bits=8` (or lower/higher) per observer instance.  
- **Outlier Handling**  
  - Consider percentile clipping (e.g. 99.9%) rather than raw min/max to reduce spikes.  
- **Module Fusion**  
  - If you have `Conv + BatchNorm`, fuse before wrapping to reduce error.  
- **Multiple Algorithms**  
  - Implement each PTQ algorithm as a separate Observer subclass. The wrapper holds a `dict` of observers and lets you pick one (or run several side-by-side).  
- **State Dict & Serialization**  
  - Ensure `state_dict()` includes both model weights and all observer parameters (running stats, final scales/zero-points).  
- **Performance & Overhead**  
  - Fake-quant calls inside `forward` will slow down calibration/inference—allow disabling them once evaluation is done.  
- **Export & Backends**  
  - Verify that your target backend supports the fake-quant ops you emit; you may need to remap them to backend-specific quant ops after export.

### Sketch of the API & Class Structure

```python
import torch
import torch.nn as nn

class ObserverBase:
    def __init__(self, dtype=torch.qint8, qscheme=torch.per_tensor_affine, num_bits=8):
        self.dtype = dtype
        self.qscheme = qscheme
        self.num_bits = num_bits
        self.enabled = True
        self.init_stats()

    def init_stats(self):
        # e.g. running min/max or histogram bins
        self.min_val = float('inf')
        self.max_val = float('-inf')

    def collect(self, x: torch.Tensor):
        if not self.enabled: return
        self.min_val = min(self.min_val, float(x.min()))
        self.max_val = max(self.max_val, float(x.max()))

    def compute_qparams(self):
        qmin, qmax = 0, 2 ** self.num_bits - 1
        scale = (self.max_val - self.min_val) / float(qmax - qmin)
        zero_point = qmin - round(self.min_val / scale)
        zero_point = int(max(qmin, min(qmax, zero_point)))
        return torch.tensor([scale]), torch.tensor([zero_point], dtype=torch.int)

    def fake_quant(self, x: torch.Tensor):
        scale, zero_point = self.compute_qparams()
        return torch.fake_quantize_per_tensor_affine(
            x, scale=scale, zero_point=zero_point,
            quant_min=0, quant_max=2**self.num_bits-1
        )

class PTQWrapper(nn.Module):
    def __init__(self, module: nn.Module, algos: dict[str, ObserverBase]):
        super().__init__()
        self.module = module
        self.algos = algos
        self.active_algo = next(iter(algos))
        self.calibration_enabled = False
        self.quant_enabled = False

    def enable_calibration(self, algo_name: str=None):
        self.calibration_enabled = True
        self.quant_enabled = False
        if algo_name:
            self.active_algo = algo_name

        for obs in self.algos.values():
            obs.enabled = True
            obs.init_stats()

    def compute_qparams(self):
        obs = self.algos[self.active_algo]
        obs.enabled = False  # stop collecting
        self.quant_enabled = True

    def forward(self, x):
        obs = self.algos[self.active_algo]

        # TODO Consider weight

        # Call the underlying module
        out = self.module(x)

        # Collect activation stats or Fake-quantize output 
        if self.calibration_enabled and obs.collect_activations:
            obs.collect(out)                       
        if self.quant_enabled:
            out = obs.fake_quant(out)

        return out
```

###  Usage Example

```python
# 1. Build your FP32 model
model = MyModel()

# 2. Wrap each layer you want to quantize
for name, module in model.named_modules():
    if isinstance(module, (nn.Conv2d, nn.Linear)):
        algos = {
            'minmax': ObserverBase(num_bits=8),
            'percentile': PercentileObserver(num_bits=8, percentile=99.9),
        }
        wrapper = PTQWrapper(module, algos)
        parent, attr = find_parent(model, name)
        setattr(parent, attr, wrapper)

# 3. Calibrate
model.eval()
for batch in calibration_dataloader:
    model.enable_calibration(algo_name='minmax')
    _ = model(batch)

# 4. Compute & freeze qparams
for module in model.modules():
    if isinstance(module, PTQWrapper):
        module.compute_qparams()

# 5. Evaluate quantized-simulation model
with torch.no_grad():
    acc = evaluate(model, val_loader)

# 6. Export
exported = torch.export.export(model, inputs=(torch.randn(1,3,224,224),))
```
